### ISSUE TITLE
special.jn_zeros(281, 6) hangs

### ISSUE THREAD

**Issue by WarrenWeckesser** [scipy/scipy#4690]
As reported at http://stackoverflow.com/questions/29415795/scipy-hangs-on-finding-zeros-of-bessel-functions, the call to `special.jn_zeros(281, 6)` hangs.  I can reproduce this with scipy versions `0.15.1` and `0.16.0.dev0+4585b0f`.  `jn_zeros(280, 6)` returns 6 values, but `jn_zeros(281, 6)` never returns, and I have to kill the python process.

I'm using python 2.7.9 and numpy 1.9.2 on a Mac.

**Comment by WarrenWeckesser** [comment]
The problem is in the subroutine `JYZO` in `special/specfun/specfun.f`.  This subroutine computes the zeros of J, J', Y and Y' using Newton's method.  (This is the function used by `jn_zeros`, `jnp_zeros`, `yn_zeros` and `ynp_zeros`.)  When the code sees that Newton's method has converged to a previously found zero of Y', it adds π to the initial guess and tries again.  Neither Newton's method nor the retry loop are protected from iterating indefinitely.  In the particular case of `jn_zeros(281, 6)`, eventually it uses an initial guess for a zero of Y' that puts Newton's method into an infinite loop.

Even if Newton's method converges to a new zero, there is no guarantee that it is actually the next zero to occur after the previously found zero.  For example, note the big gaps between the zeros returned by `ynp_zeros(m, 4)` when `m` is large:

    >>> ynp_zeros(280, 4)
    array([ 292.05733785,  394.29885128,  398.73595856,  403.12550867])
    >>> ynp_zeros(282, 4)
    array([ 294.08529796,  342.7659478 ,  353.48703413,  373.42477246])

**Comment by pv** [comment]
Proper implementation here probably requires using
http://dlmf.nist.gov/10.21#viii

05.04.2015, 18:45, Warren Weckesser kirjoitti:

**Comment by WarrenWeckesser** [comment]
I won't have time to work on a proper fix for this, but I did a little experimenting.  The zeros of J provide a pretty good approximation to the zeros of Y'.  When I changed the loop that computes the zeros of Y' to use the previously computed zeros of J as the initial guess in Newton's method, the problem went away, and there was no need to check for convergence to a previous zero (at least for the values I tried, which included the first 100 or so zeros, for orders up to 1000).

A "proper fix" might include splitting up the code so that, for example, the zeros of J, J' and Y are not computed when only the zeros of Y' are requested, in which case the above observation is not useful.
